src/xdp-utils.c

```diff
@@ -117,6 +121,198 @@ needs_quoting (const char *arg)
+{
+  size_t len;
+  char *end;
+  long n;
+
+  g_assert (value != NULL);
+  g_assert (expected_length > 0 && expected_length <= 4);
+
+  len = strlen (value);
+  if (len != expected_length)
+    return FALSE;
+
+  n = strtol (value, &end, 16);
+
+  if (end - value != len)
+    return FALSE;
+
+  if (n <= 0 || n > UINT16_MAX)
+    return FALSE;
+
+  if (out_value)
+    *out_value = n;
+
+  return TRUE;
+}
+
+static gboolean
+parse_all_usb_rule (XdpUsbRule  *dest,
+                    GStrv        data)
+{
+  if (g_strv_length (data) != 1)
+    return FALSE;
+
+  dest->rule_type = XDP_USB_RULE_TYPE_ALL;
+  return TRUE;
+}
+
+static gboolean
+parse_cls_usb_rule (XdpUsbRule  *dest,
+                    GStrv        data)
+{
+  const char *subclass;
+  const char *class;
+
+  if (g_strv_length (data) < 3)
+    return FALSE;
+
+  class = data[1];
+  subclass = data[2];
+
+  if (!validate_hex_uint16 (class, 2, &dest->d.device_class.class))
+    return FALSE;
+
+  if (g_strcmp0 (subclass, "*") == 0)
+    dest->d.device_class.type = XDP_USB_RULE_CLASS_TYPE_CLASS_ONLY;
+  else if (validate_hex_uint16 (subclass, 2, &dest->d.device_class.subclass))
+    dest->d.device_class.type = XDP_USB_RULE_CLASS_TYPE_CLASS_SUBCLASS;
+  else
+    return FALSE;
+
+  dest->rule_type = XDP_USB_RULE_TYPE_CLASS;
+  return TRUE;
+}
+
+static gboolean
+parse_dev_usb_rule (XdpUsbRule *dest,
+                    GStrv       data)
+{
+  if (g_strv_length (data) != 2)
+    return FALSE;
+
+  if (!validate_hex_uint16 (data[1], 4, &dest->d.product.id))
+    return FALSE;
+
+  dest->rule_type = XDP_USB_RULE_TYPE_DEVICE;
+  return TRUE;
+}
+
+static gboolean
+parse_vnd_usb_rule (XdpUsbRule *dest,
+                    GStrv       data)
+{
+  if (g_strv_length (data) != 2)
+    return FALSE;
+
+  if (!validate_hex_uint16 (data[1], 4, &dest->d.product.id))
+    return FALSE;
+
+  dest->rule_type = XDP_USB_RULE_TYPE_VENDOR;
+  return TRUE;
+}
+
+static const struct {
+  const char *name;
+  gboolean (*parse) (XdpUsbRule *dest,
+                     GStrv       data);
+} rule_parsers[] = {
+  { "all", parse_all_usb_rule },
+  { "cls", parse_cls_usb_rule },
+  { "dev", parse_dev_usb_rule },
+  { "vnd", parse_vnd_usb_rule },
+};
+
+static XdpUsbRule *
+xdp_usb_rule_from_string (const char *string)
+{
+  g_autoptr(XdpUsbRule) usb_rule = NULL;
+  g_auto(GStrv) split = NULL;
+  gboolean parsed = FALSE;
+
+  split = g_strsplit (string, ":", 0);
+
+  if (!split || g_strv_length (split) > 3)
+    return NULL;
+
+  usb_rule = g_new0 (XdpUsbRule, 1);
+
+  for (size_t i = 0; i < G_N_ELEMENTS (rule_parsers); i++)
+    {
+      if (g_strcmp0 (rule_parsers[i].name, split[0]) == 0)
+        {
+          if (!rule_parsers[i].parse (usb_rule, split))
+            return FALSE;
+
+          parsed = TRUE;
+          break;
+        }
+    }
+
+  if (!parsed)
+    return NULL;
+
+  return g_steal_pointer (&usb_rule);
+}
+
+void
+xdp_usb_query_free (XdpUsbQuery *query)
+{
+  if (!query)
```

Comment:
Same here

Reply:
no, we dereference query.

Reply:
Don't pass in invalid pointers then. Functions can have pre-conditions.

Reply:
when pre-conditions are NOT checked with any tool then they are just footguns. I know the tooling suck here, it's not a justification for dangerous practices. It should be safe with a NULL pointer.

Reply:
it's now g_return_if_fail()